src/bn_balances.erl

```diff
@@ -0,0 +1,195 @@
+    lager:notice("Unhandled info ~p", [Info]),
+    {noreply, State}.
+
+terminate(_Reason, #state{db = DB}) ->
+    rocksdb:close(DB).
+
+%%
+%% Hooks
+%%
+
+incremental_commit_hook(Changes) -> 
+    {ok, #state{db=DB, default=DefaultCF, entries=EntriesCF, entry_heights=EntryHeightsCF}} = get_state(),
+    Chain = blockchain_worker:blockchain(),
+    {ok, Height} = blockchain:height(Chain),
+    {ok, Batch} = rocksdb:batch(),
+    %% Load initial balances for all accounts before adding new commits
+    case rocksdb:get(DB, <<"snapshot_balances_loaded">>, []) of
+        not_found ->
+            {ok, CurrentLedger} = blockchain:ledger_at(Height, Chain),
+            EntriesBin = blockchain_ledger_v1:snapshot_raw_accounts(CurrentLedger),
+            lager:info("Loading initial balances at snapshot height ~p for ~p accounts.", [Height, length(EntriesBin)]),
+            lists:foreach(
+                fun({AddressBin, EntryBin}) ->
+                    Address = ?BIN_TO_B58(AddressBin),
+                    EntryKeyBin = erlang:term_to_binary({Address, Height}),
+                    % Store initial height for all entries at snapshot height
+                    EntryHeightsList = case rocksdb:get(DB, EntryHeightsCF, AddressBin, []) of
+                        not_found ->
+                            [];
+                        {error, Error} ->
+                            {error, Error};
+                        {ok, EntryHeightsListBin} ->
+                            erlang:binary_to_term(EntryHeightsListBin)
+                    end,
+                    NewEntryHeightsList = lists:append(EntryHeightsList, [Height]),
+                    NewEntryHeightsListBin = erlang:term_to_binary(NewEntryHeightsList),
+                    rocksdb:batch_put(Batch, EntriesCF, EntryKeyBin, EntryBin),
+                    rocksdb:batch_put(Batch, EntryHeightsCF, AddressBin, NewEntryHeightsListBin)
+                end,
+                EntriesBin
+            ),
+            rocksdb:batch_put(Batch, DefaultCF, <<"snapshot_balances_loaded">>, <<"true">>),
+            rocksdb:write_batch(DB, Batch, []);
+        {ok, _} ->
+            ok;
+        {error, InitError} ->
+            lager:error("Error getting snapshot_balances_loaded from DB ~p: ~p", [DB, InitError])
+    end,
+    lager:info("Saving ~p entry changes", [length(Changes)]),
+    lists:foldl(
+        fun(Change, _) ->
+            {_, _, AddressBin, EntryBin} = Change,
+            % New entry heights are stored in a list for each address in the EntryHeightsCF
+            EntryHeightsList = case rocksdb:get(DB, EntryHeightsCF, AddressBin, []) of
+                not_found ->
+                    [];
+                {error, Error} ->
+                    {error, Error};
+                {ok, EntryHeightsListBin} ->
+                    erlang:binary_to_term(EntryHeightsListBin)
+            end,
+            NewEntryHeightsList = lists:append(EntryHeightsList, [Height]),
```

Comment:
Would it be better just to iterate backwards on all entries? I'm not sure the implication of having a large value vs the lookup time for a large balances.db if it's a long running node.

Maybe i could just store the total number of entries for an address instead of a list of heights, and use a nested value that includes height {Address, EntryIndex} => { height => X, entry => Y}

open to other ideas. I'm trying to avoid just iterating over the entire balances.db... though if for some reason that's the better path i'm willing to give it a shot.